tools/system_message.toit

```diff
@@ -42,52 +47,109 @@ handle-system-message encoded-system-message snapshot-content -> none
+            --arguments="WyNVBVVYU...VQhJBc0=",
+        cli.Example "Decode a message, using the supplied snapshot:"
+            --arguments="--snapshot foo.snapshot WyNVBVVQ...VQhJBc0=",
+      ]
+  return cli.Command "decode"
       --help="""
-        Decodes system messages from devices.
+        Decodes system messages.
 
         System messages encode stack traces, profile runs, and other information.
-        This utility is automatically called  by `jag decode` to provide nice
+        This utility is automatically called  by `jag monitor` to provide nice
         output from the encoded messages a device prints on the serial port.
+
+        Searches for a snapshot file based on the UUID of the program that produced
+        the message. If the snapshot file is not found, the message is still decoded
+        but the output will be less informative.
+
+        Searches in the following directories for snapshot files:
+        $((search-dirs.map: "- $it").join "\n")
         """
-      --options=[
-        cli.Option "snapshot" --short-name="s"
-            --help="The snapshot file of the program that produced the message",
-        cli.Option "message" --short-name="m" --required
-            --help="The base64-encoded message from the device",
-        cli.Option "uuid" --short-name="u"
-            --help="UUID of the snapshot that produced the message",
-        cli.Flag "force-pretty"
-            --help="Force the report to use terminal graphics",
-        cli.Flag "force-plain"
-            --help="Force the report to be pure ASCII even on a terminal",
-      ]
-      --run=:: decode-system-message it command
+      --options=options
+      --rest=rest
+      --examples=examples
+      --run=:: decode it
+
+// TODO(florian): when removing this entry-point, rename the file and make the
+// "system-message_.toit" file the new "system-message.toit".
+// Also remove the "use-old-api" flag from the "build-command" function.
+main args:
+  command := build-command --use-old-api
   command.run args
 
-decode-system-message parsed command -> none:
+compute-snapshot-dirs -> List:
+  result := []
+  result.add (fs.join xdg.state-home "toit" "snapshots")
+  result.add (fs.join xdg.cache-home "jaguar" "snapshots")
+  return result
+
+decode parsed/cli.Parsed -> none:
   if not parsed: exit 1
+
+  if parsed["uuid"]:
+    pipe.print-to-stdout "The --uuid flag is deprecated and will be removed in a future release."
+
+  // The encoded-system-message is ubjson-encoded.
   encoded-system-message := base64.decode parsed["message"]
+  // The decoded message still has its payload encoded in JSON.
+  decoded-system-message := decode-system-message encoded-system-message --on-error=:
+    pipe.print-to-stderr it
+    exit 1
+
+  snapshot-path := parsed["snapshot"]
+  if not snapshot-path:
+    // Try to find the snapshot based on the UUID.
+    dirs := compute-snapshot-dirs
```

Comment:
Not sure if this is the best way of factoring this. Maybe it would be nicer if you just had a function to find the snapshot-path and inlined the compute-snaphots-dir into that. Then you could use return instead of break from there.

Reply:
Still need the `compute-snapshot-dirs` for the help, but created a function.
done.